### ISSUE TITLE
`cacheFrom` with a reference to uncreated ECR repository causes error: could not open dockerfile at relative path Dockerfile

### ISSUE THREAD

**Issue by jchook** [pulumi/pulumi-docker#676]
### What happened?

My goal is:

* Create an ECR repository
* Create an Image resource and push it to that repository
  * Specify some meaningful value for cacheFrom that references the repository resource details (such as it's URL).

When I input a `cacheFrom` that references the ECR repository resource via `pulumi.all`, I get a strange unrelated error.

> error: could not open dockerfile at relative path Dockerfile

From what I can guess, it appears a squelched internal error caused by the reference to the uncreated repository totally zaps/resets the Image build inputs. I can somewhat verify this by running `cp app/Dockerfile .` in the minimal repro.

```
     Type                   Name                   Plan       Info
 +   pulumi:pulumi:Stack    pulumi-docker-bug-dev  create
 +   ├─ aws:ecr:Repository  app                    create
     └─ docker:index:Image  app                               1 error


Diagnostics:
  docker:index:Image (app):
    error: could not open dockerfile at relative path Dockerfile: stat Dockerfile: no such file or directory
```

### Expected Behavior

On the initial deployment, the `cacheFrom.images` Input using `pulumi.all` resolves to an empty array and the build carries on as normal.

### Steps to reproduce

See a [minimal reproduction of the issue here](https://github.com/jchook/pulumi-issue-676)

index.ts

```typescript
import * as pulumi from '@pulumi/pulumi'
import * as aws from '@pulumi/aws'
import * as docker from '@pulumi/docker'

const myName = 'app'
const repository = new aws.ecr.Repository(myName, {
  name: myName,
  tags: {
    Name: myName,
  },
})

const authToken = aws.ecr.getAuthorizationTokenOutput({
  registryId: repository.registryId,
})

const image = new docker.Image(myName, {
  build: {
    args: {
      BUILDKIT_INLINE_CACHE: '1',
    },

    // ERROR!
    //
    // Adding cacheFrom with a reference to `repository` causes this error:
    // error: could not open dockerfile at relative path Dockerfile
    //
    // The error is not present if:
    // - cacheFrom input does not reference an uncreated repository
    // - the repository already exists in AWS
    //
    cacheFrom: {
      images: pulumi
        .all([repository.repositoryUrl])
        .apply(([url]) => {
          // I do not see output from this console.log
          console.log('cacheFrom repositoryUrl URL:', url)
          return []
        }),
    },
    context: './app',
  },
  imageName: pulumi.interpolate`${repository.repositoryUrl}:latest`,
  registry: {
    password: pulumi.secret(authToken.password),
    server: repository.repositoryUrl,
    username: authToken.userName,
  },
}, { dependsOn: [repository] })

export const app = {
  authToken,
  image,
  repository,
}
```

### Output of `pulumi about`

```
CLI          
Version      3.73.0
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.41.0
awsx    1.0.2
docker  4.3.0
docker  3.6.1
nodejs  unknown

Host     
OS       void
Version  
Arch     x86_64

This project is written in nodejs: executable='/home/jchook/.local/opt/n/bin/node' version='v18.16.0'

Current Stack: jchook/pulumi-docker-bug/dev

Found no resources associated with dev

Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/jchook
User           jchook
Organizations  jchook

Dependencies:
NAME            VERSION
@pulumi/aws     5.41.0
@pulumi/awsx    1.0.2
@pulumi/docker  4.3.0
@pulumi/pulumi  3.74.0
@types/node     16.18.38

Pulumi locates its logs in /tmp by default
```

### Additional context

Am I supposed to include the repository URL in the cacheFrom images list?

Possibly related: https://github.com/pulumi/pulumi-docker/issues/501

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by AaronFriel** [comment]
Hey @jchook, thanks for the repro - I was able to verify this on my machine as well. Not quite the same issue as #501, but similar.

It looks like as a workaround, you can use `pulumi up --skip-preview`, or ensure that the image is created in a second `up` after creating the repository first. We recognize that's not ideal, so I'm keeping this open as a bug for us to address.

**Comment by guineveresaenger** [comment]
I think this may be a dupe of #620.

**Comment by ossareh** [comment]
I was just burned by this for a few hours, thank you for the report - it was instrumental in stopping me from losing more hours to this bug 😅 The Pulumi AI was also very confused by my findings 🤦🏻‍♂️

May I suggest annotating the documentation to make it clear this isn't working currently?

https://www.pulumi.com/registry/packages/docker/api-docs/image/#docker-image-build-using-caching-with-aws-elastic-container-registry

**Comment by ossareh** [comment]
> It appears that if there are any dynamic properties on the build input, then the provider erroneously resets the build object to use the default Dockerfile location.

I tested this assertion; this is what I found.

Attempt 1:
* `const repo = new aws.ecr.Repository(...)` # this is a new, not yet existing repo
* `const token = aws.ecr.getAuthroizatonTokenOutput(...)`
* `new docker.Image(...)` referencing:
   * `repo` in `build.cacheFrom.images[]` and `repo.registry.server`
   * `token.password` in `registry.server.passwword`
   
 Failed with the error this ticket describes.
 
 Attempt 2:
 * `const repo = new aws.ecr.Repository(...)` # this is a new, not yet existing repo
 
Succeeds in creating `repo`

Attempt 3:
 * `const repo = new aws.ecr.Repository(...)` # this now exists
 * `const token = aws.ecr.getAuthroizatonTokenOutput(...)`
 * `new docker.Image(...)` referencing:
    * `token.password` in `registry.server.passwword`
    * hard coded ecr repo url, etc
    
This works; finds the correct Dockerfile, builds, etc. **BUT** fails to push w/ error as below.

Attempt 4:
 * `const repo = new aws.ecr.Repository(...)` # this now exists
 * `const token = aws.ecr.getAuthroizatonTokenOutput(...)`
 * `new docker.Image(...)` referencing:
    * `token.password` in `registry.server.passwword`
   * `repo` in `repo.registry.server`
   * hard coded `cacheFrom.images[]`

This works; finds the correct Dockerfile, builds, etc. **BUT** fails to push w/ error as below.

Attempt 5:
 * `const repo = new aws.ecr.Repository(...)` # this now exists
 * `const token = aws.ecr.getAuthroizatonTokenOutput(...)`
 * `new docker.Image(...)` referencing:
    * `token.password` in `registry.server.passwword`
   * `repo` in `build.cacheFrom.images[]` and `repo.registry.server`

This works; finds the correct Dockerfile, builds, etc. **BUT** fails to push w/ error as below.

Attempt 6:
 * `const repo = new aws.ecr.Repository(...)` # this now exists
 * `const token = aws.ecr.getAuthroizatonTokenOutput(...)`
 * `new docker.Image(...)` referencing:
    * `token.password` in `registry.server.passwword`
    * `token.userName` in `registry.server.username`
   * `repo` in `build.cacheFrom.images[]` and `repo.registry.server`

This works; full end to end.

So the primary take away; you cannot use dynamic attributes for repos that don't exist yet - that causes the error.

The example should be updated to show setting the username is required; otherwise you get 

```
  Diagnostics:
    docker:index:Image (software-001-development-image):
      Building your image for linux/amd64 architecture.
      To ensure you are building for the correct platform, consider explicitly setting the `platform` field on ImageBuildOptions.
      warning: username was not set, although password was; using host credentials file
      warning: username was not set, although password was; using host credentials file
      error: no basic auth credentials
```

**Comment by jchook** [comment]
> So the primary take away; you cannot use dynamic attributes for repos that don't exist yet - that causes the error.

Yes. This is well said.

The internal error appears to be squelched as it doesn't appear in debug output.

**Comment by mikhailshilkov** [comment]
I closed #787 as a duplicate, but copying my notes here:

For engineers: The cause for it is that the `build` property comes as a single unknown

```
{"build":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","imageName":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","registry":{"password":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","server":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","username":"04da6b54-80e4-46f7-96ec-b56ff0331ba9"},"skipPush":false}
```

instead of a map with a few unknown properties (that would be similar to `registry above, not yet sure why the difference). 

We then discard unknowns [here](https://github.com/pulumi/pulumi-docker/blob/24161acd0433b061e3b5acfa128b9ddb599b965b/provider/provider.go#L131) and proceed with filling default build properties in [`marshalBuildAndApplyDefaults`](https://github.com/pulumi/pulumi-docker/blob/24161acd0433b061e3b5acfa128b9ddb599b965b/provider/provider.go#L139). 

This is the same problem as https://github.com/pulumi/pulumi-docker/issues/620.

**Comment by mikhailshilkov** [comment]
Hmm, maybe this is different enough from #620 to keep it open. Sorry for the noise.
